### PR TITLE
Update <releases> in appdata

### LINKED ISSUE #17: Support updating <releases> in appdata
Flathub now strongly encourages including release information in appdata; it is a hard requirement for PRs, and someday will be required for real builds, too. So this tool will have to learn to update that in cases where the appdata is not provided by upstream.

There are a number of ways we might determine the version number:

* Some packages include the version number in their URL (or are Debian packages with a structured version number)
* Some packages might have entries on release-monitoring.org which includes version information #16
* AppImages can be unpacked by passing `--appimage-extract`; the `.desktop` file within includes an `X-AppImage-Version` key

### PR DESCRIPTION
**Pull request by wjt**
Doing this without generating a huge diff was a lot more involved than I
expected, but this seems to work OK now.

I have tested this locally against:

* endlessm/eos-google-chrome-app, which currently has no `<releases>`: the checker adds it
* flathub/us.zoom.Zoom, which already has `<releases>`: it prepends a new one, albeit with the attributes in a different order to the existing entries. There's a test marked `expectedFailure` for this; the order has no semantic impact and the 18 repos I checked have no trend towards one or another.

Fixes #17.

### CODE REVIEW

File: src/checker.py
Comment (andrunko) [inline src/checker.py]: hmm, ideally we should have a way to get the latest release info (version+date) for a "top" manifest (from path) instead, but while we don't support that I think using the last changed source is Ok, although I can imagine a situation where we have multiple external data/sources using different versions.

On possible option would be to grow support for specialized manifest checkers (alongside extra-data checkers), with each manifest checker running the external data checkers for that component + providing release info.

File: src/lib/appdata.py
Comment (andrunko) [inline src/lib/appdata.py:24]: nit: missing newline

File: src/lib/appdata.py
Comment (andrunko) [inline src/lib/appdata.py:20]: nit: missing newline

File: src/lib/appdata.py
Comment (andrunko) [inline src/lib/appdata.py]: Not sure this snippet adds anything here, maybe something for a commit msg instead?

File: src/lib/appdata.py
Comment (andrunko) [inline src/lib/appdata.py]: s/data/date

File: src/lib/appdata.py
Comment (andrunko) [inline src/lib/appdata.py]: I would probably s/in_/path/ here as well, or even `appdata_path`.

File: tests/test_appdata.py
Comment (andrunko) [inline tests/test_appdata.py:19]: nit: missing newline

**Review by andrunko — CHANGES_REQUESTED** [review]

File: src/checker.py
Comment (wjt) [reply src/checker.py]: Agreed. I realised that `date.today()` is actually not good enough: the branch name used depends on the contents of the tree, so it would change every day and file a duplicate PR until one is merged. Unfortunately the Debian repo format doesn't include timestamps, so I make a `HEAD` request against the package URL and use its timestamp.

File: src/lib/appdata.py
Comment (wjt) [reply src/lib/appdata.py]: Moved to commit messaeg.

File: src/lib/appdata.py
Comment (wjt) [reply src/lib/appdata.py]: Actually it doesn't have to be a path. It can be a file-like object too, and the test suite relies on this. I've made this explicit in the name.

File: src/checker.py
Comment (andrunko) [reply src/checker.py]: Another option would be to extract the date from the package changelog itself (which may require fetching/extracting the package if python apt doesn't provide that) and for firefox we could get the date from `LAST_RELEASE_DATE` returned on https://product-details.mozilla.org/1.0/firefox_versions.json.

**Review by andrunko — APPROVED** [review]
lgtm with a small comment regarding the release date/timestamps for debian packages and firefox. I will leave it to you to decide whether you want to work on it (I am fine to do it as a follow-up change).

File: src/checker.py
Comment (wjt) [reply src/checker.py]: `python-apt` has some support for fetching changelogs for Debian and Ubuntu, but we don't have a changelog server. https://salsa.debian.org/apt-team/python-apt/blob/master/apt/package.py#L1245-1417

The other option would be to fetch the package (or just the relevant half of the source package – if we have a source package, that is!) and parse the changelog (if present). Personally, I think we can make do with the `HEAD` request for now. I'll add a TODO about this.

---

https://product-details.mozilla.org/1.0/firefox_versions.json has `"LAST_RELEASE_DATE": "2019-07-09"` but https://www.mozilla.org/en-US/firefox/68.0.2/releasenotes/ says:

> Version 68.0.2, first offered to Release channel users on August 14, 2019

So I don't think that's going to fly. (I'll add a note about this.) However, we can and should propagate the version number from that file into the `ExternalFile` object. I'll make that change.

While looking at this, I realised that we currently fetch all the files for Firefox, but we don't need to – we get their paths from `SHA256SUMS` so we have those, and we can just make a `HEAD` request to get the size. That's another task though.

**Pull request merged**
